runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/status/ChannelzServlet.java

```diff
@@ -0,0 +1,245 @@
+            return;
+        }
+        visitedSets.channels.add(channelRef.getChannelId());
+        SettableFuture<GetChannelResponse> future = SettableFuture.create();
+        channelzService.getChannel(
+                GetChannelRequest.newBuilder().setChannelId(channelRef.getChannelId()).build(),
+                getStreamObserver(future));
+        Channel channel;
+        try {
+            channel = future.get().getChannel();
+        } catch (Exception e) {
+            String msg = "Failed to get Channel: " + channelRef;
+            LOG.warn(msg, e);
+            writer.println(msg + " Exception: " + e.getMessage());
+            return;
+        }
+        appendChannel(channel, writer, visitedSets);
+    }
+
+    private void appendChannel(Channel channel, PrintWriter writer, VisitedSets visitedSets) {
+        writer.println("<table border='1'>");
+        writer.println("<tr>");
+        writer.println("<td>");
+        writer.println("ChannelId: " + channel.getRef().getChannelId());
+        writer.println("</td>");
+        writer.println("<td><pre>" + channel);
+        writer.println("</pre></td>");
+        writer.println("</tr>");
+        appendChannels(channel.getChannelRefList(), writer, visitedSets);
+        appendSubChannels(channel.getSubchannelRefList(), writer, visitedSets);
+        appendSockets(channel.getSocketRefList(), writer);
+        writer.println("</table>");
+    }
+
+    private void appendSubChannels(List<SubchannelRef> subchannelRefList, PrintWriter writer, VisitedSets visitedSets) {
+        for (SubchannelRef subchannelRef : subchannelRefList) {
+            writer.println("<tr>");
+            writer.println("<td>");
+            writer.println("Sub Channel: " + subchannelRef.getSubchannelId());
+            writer.println("</td>");
+            writer.println("<td>");
+            appendSubchannel(subchannelRef, writer, visitedSets);
+            writer.println("</td>");
+            writer.println("</tr>");
+        }
+    }
+
+    private void appendSubchannel(SubchannelRef subchannelRef, PrintWriter writer, VisitedSets visitedSets) {
+        if (visitedSets.subchannels.contains(subchannelRef.getSubchannelId())) {
+            String msg = "Duplicate Subchannel Id: " + subchannelRef;
+            LOG.warn(msg);
+            writer.println(msg);
+            return;
+        }
+        visitedSets.subchannels.add(subchannelRef.getSubchannelId());
+        SettableFuture<GetSubchannelResponse> future = SettableFuture.create();
+        channelzService.getSubchannel(
+                GetSubchannelRequest.newBuilder().setSubchannelId(subchannelRef.getSubchannelId())
+                        .build(),
+                getStreamObserver(future));
+        Subchannel subchannel;
+        try {
+            subchannel = future.get().getSubchannel();
+        } catch (Exception e) {
+            String msg = "Failed to get Subchannel: " + subchannelRef;
+            LOG.warn(msg, e);
```

Comment:
not sure it's worth logging, can just show it in the output page

ditto for others

Reply:
done